homeassistant/components/openrgb/coordinator.py

```diff
@@ -0,0 +1,178 @@
+                    },
+                ) from err
+
+        # Return devices indexed by their key
+        return {self._get_device_key(device): device for device in self.client.devices}
+
+    def _client_update(self) -> None:
+        try:
+            self.client.update()
+        except CONNECTION_ERRORS:
+            # Try to reconnect once
+            self.client.disconnect()
+            self.client.connect()
+            self.client.update()
+
+    def _get_device_key(self, device: Device) -> str:
+        """Build a stable device key.
+
+        Note: the OpenRGB device.id is intentionally not used because it is just
+        a positional index that can change when devices are added or removed.
+        """
+        parts = (
+            self.entry_id,
+            device.type.name,
+            device.metadata.vendor or "none",
+            device.metadata.description or "none",
+            device.metadata.serial or "none",
+            device.metadata.location or "none",
+        )
+        # Double pipe is readable and is unlikely to appear in metadata
+        return "||".join(parts)
+
+    async def async_client_disconnect(self, *args) -> None:
+        """Disconnect the OpenRGB client."""
+        if not hasattr(self, "client"):
+            # Just in case
+            return
+        async with self._client_lock:
+            await self.hass.async_add_executor_job(self.client.disconnect)
+
+    def get_client_protocol_version(self) -> str:
+        """Get the OpenRGB client protocol version."""
+        return f"{self.client.protocol_version} (Protocol)"
+
+    async def async_device_set_color(self, device: Device, color: RGBColor) -> None:
+        """Set the color of a device."""
+        async with self._client_lock:
+            try:
+                await self.hass.async_add_executor_job(device.set_color, color, True)
+            except (*CONNECTION_ERRORS, ValueError) as err:
+                raise HomeAssistantError(
+                    translation_domain=DOMAIN,
+                    translation_key="failed_to_set_color",
+                    translation_placeholders={
+                        "server_address": self.server_address,
+                        "error": str(err),
+                    },
+                ) from err
+
+    async def async_device_set_mode(self, device: Device, mode: str) -> None:
+        """Set the mode of a device."""
+        async with self._client_lock:
+            try:
+                await self.hass.async_add_executor_job(device.set_mode, mode)
+            except (*CONNECTION_ERRORS, ValueError) as err:
+                raise HomeAssistantError(
+                    translation_domain=DOMAIN,
+                    translation_key="failed_to_set_mode",
+                    translation_placeholders={
+                        "server_address": self.server_address,
+                        "error": str(err),
+                    },
+                ) from err
```

Comment:
Looking at extensiblity, I would argue that we should make the lock a public variable and move this to the platform where we use it